tests/tracking/test_rest_tracking.py

```diff
@@ -270,6 +274,25 @@ def test_log_metrics_params_tags(mlflow_client, backend_store_uri):
     mlflow_client.set_tag(run_id, "taggity", "do-dah")
     run = mlflow_client.get_run(run_id)
     assert run.data.metrics.get("metric") == 123.456
+
+    with pytest.raises(MlflowException, match=r"Invalid value for 'run_id' supplied"):
+        mlflow_client.log_metric(31, key="metric", value=41, timestamp=59, step=26)
+
+    with pytest.raises(MlflowException, match=r"Invalid value for 'key' supplied"):
+        mlflow_client.log_metric(run_id, key=31, value=41, timestamp=59, step=26)
+
+    with pytest.raises(MlflowException, match=r"Invalid value for 'value' supplied"):
+        mlflow_client.log_metric(run_id, key="foo", value=True, timestamp=31, step=41)
+
+    with pytest.raises(MlflowException, match=r"Invalid value for 'run_id' supplied"):
+        mlflow_client.log_param(31, key="foo", value="bar", timestamp=59, step=26)
+
+    with pytest.raises(MlflowException, match=r"Invalid value for 'key' supplied"):
+        mlflow_client.log_param(run_id, key=31, value="bar", timestamp=59, step=26)
+
+    with pytest.raises(MlflowException, match=r"Invalid value for 'value' supplied"):
```

Comment:
Can we add test cases for required fields and thresholds, as well as list-typed fields?

Reply:
Small thing on this: It seems like proto parsing might be automatically converting singletons to lists (e.g. `None` --> `[None]`) which then fails downstream. On one hand, I could always validate on the list types with `_assert_array` to fix this issue, but that'd potentially break downstream clients who are specifying a `experiment_ids` as a single `experiment_id` for example.

Any thoughts on this one? One option would be to just not test it... I can't think of something where wrapping a singleton in a list would break, which is the only way that I can test throwing `400` for something that should be a list

Reply:
And another: If I specify something like `metrics="foo"` in `log_batch`, I get an `AttributeError` downstream:

```python
> metric_protos = [metric.to_proto() for metric in metrics]
E   AttributeError: 'str' object has no attribute 'to_proto'
```

Unsure what to do about this one either. I think these are relics of the same top level issue, which is that assuming `parse_dict` not throwing an error meaning that the types are okay isn't actually a safe assumption. 

I also have other tests failing on (e.g.) `create_experiment`, where I specify `name=3141` and it throws a type error downstream. Can you repro that issue? I tried rebasing, but maybe I'm not caught up with `master`

Reply:
@mrkaye97 These are instances of client-side exceptions thrown before the client ever sends a request to the MLflow server; here's the stack trace for the `create_experiment` scenario:

```
Traceback (most recent call last):
  File "log.py", line 5, in <module>
    mlflow.create_experiment(name=3141)
  File "/Users/corey.zumar/mlflow/mlflow/tracking/fluent.py", line 1023, in create_experiment
    return MlflowClient().create_experiment(name, artifact_location, tags)
  File "/Users/corey.zumar/mlflow/mlflow/tracking/client.py", line 507, in create_experiment
    return self._tracking_client.create_experiment(name, artifact_location, tags)
  File "/Users/corey.zumar/mlflow/mlflow/tracking/_tracking_service/client.py", line 176, in create_experiment
    return self.store.create_experiment(
  File "/Users/corey.zumar/mlflow/mlflow/store/tracking/rest_store.py", line 97, in create_experiment
    CreateExperiment(name=name, artifact_location=artifact_location, tags=tag_protos)
TypeError: 3141 has type int, but expected one of: bytes, unicode
```

Similarly, here's the stack trace for the `log_batch` scenario:

```
  File "log.py", line 8, in <module>
    MlflowClient().log_batch("123", metrics="foo")
  File "/Users/corey.zumar/mlflow/mlflow/tracking/client.py", line 918, in log_batch
    self._tracking_client.log_batch(run_id, metrics, params, tags)
  File "/Users/corey.zumar/mlflow/mlflow/tracking/_tracking_service/client.py", line 292, in log_batch
    self.store.log_batch(run_id=run_id, metrics=metrics, params=params, tags=tags)
  File "/Users/corey.zumar/mlflow/mlflow/store/tracking/rest_store.py", line 303, in log_batch
    metric_protos = [metric.to_proto() for metric in metrics]
  File "/Users/corey.zumar/mlflow/mlflow/store/tracking/rest_store.py", line 303, in <listcomp>
    metric_protos = [metric.to_proto() for metric in metrics]
AttributeError: 'str' object has no attribute 'to_proto'
```

The client is trying & failing to convert the user-specified values into a JSON payload by using proto as an intermediary. If we want to improve the clarity / fidelity of the exceptions in this case, we could apply similar exception handling logic for proto conversion failures to the client. Please let me know if you're interested in taking this on as a follow-up PR.